subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -62,13 +62,14 @@ public CurrentSnapshotResult execute(IncrementalChangesContext context) {
         if (!buildCache.isEnabled()) {
             return executeWithoutCache(context);
         }
-        CacheHandler cacheHandler = context.getWork().createCacheHandler();
+        UnitOfWork work = context.getWork();
+        CacheHandler cacheHandler = work.createCacheHandler();
         return cacheHandler
-            .load(cacheKey -> load(context.getWork(), cacheKey))
-            .map(cacheHit -> cacheHit
-                .map(loadResult -> {
-                    OriginMetadata originMetadata = loadResult.getOriginMetadata();
-                    ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = loadResult.getResultingSnapshots();
+            .load(cacheKey -> Try.ofFailable(() -> buildCache.load(commandFactory.createLoad(cacheKey, work))))
```

Comment:
I wonder if this would be better:
```java
Try.ofFailable(() -> cacheHandler.load(cacheKey ->
    buildCache.load(commandFactory.createLoad(cacheKey, work))
))
```

Then we don't need to have the `Try` in the signature of the `CacheHandler.load` command.

On the other hand, having the try in the function which loads makes clear that we handle a failure to load.

WDYT?

Reply:
I think what you recommend is a better solution, and it still manages to capture that it's the load that's fallible.